### ISSUE TITLE
Instable/Incorrect CSS import order

### ISSUE THREAD

**Issue by Netail** [vercel/next.js#62913]
### Link to the code that reproduces this issue

https://github.com/Netail/app-dir-css-order

### To Reproduce

1. Override styling through the className prop from an external component (In this case from a package within the monorepo)
2. Start the development server `yarn dev`
3. Open the root page `/`, observe; the component uses the initial CSS _(Incorrect order)_
4. Reload the root page `/`, observe; the component uses the locally overwritten CSS _(Correct order)_
5. Open the subpath `/bug`, observe; the component uses the initial CSS _(Incorrect order)_
6. Reload the subpath `/bug`, observe; the component still uses the initial CSS _(Incorrect order)_

(When the dev server is restarted, it can be reproduced the other way around - if you open the subpath first, the CSS on this page will be correct after a page reload while the other page stays broken at all times)

### Current vs. Expected behavior

Currently, the order of the classes in the bundled CSS file is not consistent, resulting in an inconsistency of style overwriting and not respecting the correct order. Only the first rendered page will have the correct CSS order (but only after a page reload), while the CSS of the other pages continue with the incorrect order.

Override (local CSS) is placed before the initial CSS
![image](https://github.com/vercel/next.js/assets/11695769/3ff211bc-7aa1-4f83-a67f-e5a94f1abd24)

All pages should maintain the correct import order from the start (not even requiring a reload)

### Provide environment information

```bash
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 10 Home
Binaries:
  Node: 20.11.1
  npm: N/A
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 14.1.2
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router

### Which stage(s) are affected? (Select all that apply)

next dev (local), next build (local)

**Comment by Netail** [comment]
Still seems to be an issue with version [14.2.0-canary.4](https://github.com/vercel/next.js/commit/1ad3963d4448ee503accfdd01c61d8860016ab42) after the merge of [#62927](https://github.com/vercel/next.js/pull/62927)

**Comment by Netail** [comment]
On MacOS you have to switch between routes to trigger the bug, instead of a simpel reload

**Comment by Netail** [comment]
This is apparently already a bug since v13...

**Comment by Netail** [comment]
Also reproducible on macOS, but slightly different (You need to open a different page first);

1. Open the root page `/`, observe; the component uses the initial CSS _(Incorrect order)_
2. Open the subpath `/bug`, observe; the component uses the initial CSS _(Incorrect order)_
3. Reopen the root page `/`, observe; the component uses the locally overwritten CSS _(Correct order)_
4. Reopen the subpath `/bug`, observe; the component still uses the initial CSS _(Incorrect order)_

**Comment by Netail** [comment]
Seems to be resolved since; 14.2.0-canary.28
